### DISCUSSION TITLE
How does JZ and test work?

### DISCUSSION THREAD

**Discussion by ghost** [community/community#109943] [Programming Help]
### Body

```asm
section .text
  global _CheckForYAxisCollision
  global _CheckPlayerLives
  global _GameOver
  global _CheckForXAxisCollision

section .data 
  PlayerYAxis: dd 5 
  playerLives: dd 10
  playerXAxis: dd 5

_CheckForYAxisCollision:
    .include "EnemyShip.asm";
    test dword [playerYAxis], dword [enemyYAxis]
    jz _CheckPlayerLives

_CheckForXAxisCollision:
    .include "EnemyShip.asm"
    test dword [playerXAxis], dword [enemyXAxis]
    jz _CheckPlayerLives

_CheckPlayerLives:
    .include "GameOverScreen.asm"
    extern _GameOverScreen
    test dword [playerLives], 0  
    jz _GameOverScreen
```

### Guidelines

- [X] I have read and understood this category's [guidelines](https://github.com/orgs/community/discussions/51384) before making this post.

**Comment by richdentistboy** [comment]
Hey there PyDOS8,

It looks like you're delving into some assembly language with your code snippet that you've provided so, let's break down what's happening:

test is used to perform a bitwise AND operation between the values stored in memory locations [playerYAxis], [playerXAxis], and [playerLives] and the values stored in [enemyYAxis] and [enemyXAxis]. This helps in determining if there's a collision or if the player has any lives left.

jz or "Jump if Zero" is a conditional jump instruction. It checks if the Zero Flag (ZF) is set, indicating that the result of the previous test operation was zero. If so, it jumps to the specified label _CheckPlayerLives, which presumably handles what happens when there's a collision or when the player runs out of lives.

So, in essence, your code is checking for collisions in the y-axis and x-axis, and then managing player lives accordingly.

Hope this sheds some light for your code, and it's helpful to you.

Best regards,
Sturmgeisty
